### DISCUSSION TITLE
How do I build and test changes?

### DISCUSSION THREAD

**Discussion by cgagliardi** [samsonmking/epaper.js#95] [Q&A]
I'm working on an implementation for [Waveshare 10.3](https://github.com/samsonmking/epaper.js/issues/84). I think I have it ready, but I can't figure out how to actually test it.

My guess was to run 

```
pnpm install
pnpm run build
```

But it doesn't look like the C files were ever compiled. Do you have documentation on this?

**Comment by cgagliardi** [comment]
I was able to compile the C code via `node-gyp build` in the relevant packages.

Now to fix the bugs!

**Reply by SCiunczyk** [reply]
Hi there,

I'm working on implementing [Waveshare 1.54](http://www.waveshare.com/wiki/1.54inch_e-Paper_Module) on NanoPi Duo 2 on Armbian. The module/hardware works - I have modified the original sources and python modules. But I want to get rid of WiringPI and as my whole project is built on Node-RED I want to have epaper.js support.  Unfortunately I'm not fluent in JS/Node programming yet ;-).

Before I transplant my modifications to epaper.js, I need to have a working toolchain. Maybe you could help me how to build and install epaper.js properly from git clone sources? It's probably something trivial, but I'm stuck. 

I have tried running it: 
`pnpm install `(in the main epaper.js folder)
`npm install -g `(in core / cli subfolders)
no errors, but
`ejs `command returns

```
root@epdtest:~# ejs
node:internal/modules/cjs/loader:1147
  throw err;
  ^

Error: Could not find module '../dist/cli'.
Need stack:
- /root/epaper.js/packages/cli/bin/ejs-cli
    on Module._resolveFilename (node:internal/modules/cjs/loader:1144:15)
    at Module.load (node:internal/modules/cjs/loader:985:27)
    on Module.require (node:internal/modules/cjs/loader:1235:19)
    at require (node:internal/modules/helpers:176:18)
    at Object.<anonymous> (/root/epaper.js/packages/cli/bin/ejs-cli:6:1)
    at Module._compile (node:internal/modules/cjs/loader:1376:14)
    at Module.extensions..js (node:internal/modules/cjs/loader:1435:10)
    on Module.load (node:internal/modules/cjs/loader:1207:32)
    at Module.load (node:internal/modules/cjs/loader:1023:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:135:12) { }.
  Code: 'MODULE_NOT_FOUND',
  requireStack: [ '/root/epaper.js/packages/cli/bin/ejs-cli' ]
}
```

Latest - fresh Armbian 23.11.1 Jammy with Linux 6.1.63-current-sunxi (node-gyp installed as well)
